File: scala/LongestPassword.scala
//https://codility.com/demo/results/trainingS8ANJ4-GBA/
object SolutionLongestPassword {
  def solution(s: String): Int = {
    def filter(strs: List[String]): List[Int] = for(str <- strs if 1 == (str.length() & 1) && str.forall(_.isLetterOrDigit) && 1 == (str.count(_.isDigit) & 1))yield str.length()
    (-1 :: filter(s.split(" ").toList)).max
  }
}

File: python/LongestPassword.py
# https://codility.com/demo/results/trainingYXSX3P-BKV/
# Simplicity of Python? Zen of Python?
# Note: the [-1] appended to the list comprehension is absoultely necessary if there is no valid result at all in given S
# max() can't handle a list with zero element and will throw exception.
def solution(S):
    return max([len(str) for str in S.split() if len(str) & 1 and str.isalnum() and sum(c.isdigit() for c in str) & 1] + [-1])
